### ISSUE TITLE
:failure status in with-workspace and event-bus

### ISSUE THREAD

**Issue by kwladyka** [flosell/lambdacd#154]
```clojure
(defn failure-notifications [ctx]
  (let [subscription (event-bus/subscribe ctx :step-finished)
        steps-finished (event-bus/only-payload subscription)]
    (async/go-loop []
      (let [event (async/<! steps-finished)]
        (when (= :failure (get-in event [:final-result :status]))
          (slack/send-message event)))
      (recur))))
```

```clojure
(with-workspace
              (steps/clone sb-repo)
              git/list-changes
              steps/lein-test)
```

When `lein-test` fail it triggers also `clone` with `:failure` status. Twice. I was trying also with `:step-result-updated`. Am i doing something wrong or is it a bug? I want this even only for step which really fail. In that case `lein-test`.

I got this messages:

slack message: 
```edn
{:final-result
 {:exit 1
  :out
  "\nlein test foo.core-test\n\nlein test :only foo.core-test/a-test\n\nFAIL in (a-test) (core_test.clj:7)\nFIXME, I fail.\nexpected: 0\n  actual: 1\n    diff: - 0\n          + 1\n\nRan 1 tests containing 1 assertions.\n1 failures, 0 errors.\nTests failed.\n"
  :status :failure}
 :rerun-for-retrigger false
 :build-number 8
 :step-id (3 1 2)}
```

slack message:  
```edn
{:final-result
 {:exit 1
  :global {}
  :out
  "Cloning git@github.com:kwladyka/lambdacd-test.git ...\n\rremote:Counting objects: 71\rremote:Counting objects: 71\n\rremote:Compressing objects: 2% ( 1/37)\rremote:Compressing objects: 5% ( 2/37)\rremote:Compressing objects: 8% ( 3/37)\rremote:Compressing objects: 10% ( 4/37)\rremote:Compressing objects: 13% ( 5/37)\rremote:Compressing objects: 16% ( 6/37)\rremote:Compressing objects: 18% ( 7/37)\rremote:Compressing objects: 21% ( 8/37)\rremote:Compressing objects: 24% ( 9/37)\rremote:Compressing objects: 27% (10/37)\rremote:Compressing objects: 29% (11/37)\rremote:Compressing objects: 32% (12/37)\rremote:Compressing objects: 35% (13/37)\rremote:Compressing objects: 37% (14/37)\rremote:Compressing objects: 40% (15/37)\rremote:Compressing objects: 43% (16/37)\rremote:Compressing objects: 45% (17/37)\rremote:Compressing objects: 48% (18/37)\rremote:Compressing objects: 51% (19/37)\rremote:Compressing objects: 54% (20/37)\rremote:Compressing objects: 56% (21/37)\rremote:Compressing objects: 59% (22/37)\rremote:Compressing objects: 62% (23/37)\rremote:Compressing objects: 64% (24/37)\rremote:Compressing objects: 67% (25/37)\rremote:Compressing objects: 70% (26/37)\rremote:Compressing objects: 72% (27/37)\rremote:Compressing objects: 75% (28/37)\rremote:Compressing objects: 78% (29/37)\rremote:Compressing objects: 81% (30/37)\rremote:Compressing objects: 83% (31/37)\rremote:Compressing objects: 86% (32/37)\rremote:Compressing objects: 89% (33/37)\rremote:Compressing objects: 91% (34/37)\rremote:Compressing objects: 94% (35/37)\rremote:Compressing objects: 97% (36/37)\rremote:Compressing objects: 100% (37/37)\rremote:Compressing objects: 100% (37/37)\n\rReceiving objects: 1% ( 1/71)\rReceiving objects: 2% ( 2/71)\rReceiving objects: 4% ( 3/71)\rReceiving objects: 5% ( 4/71)\rReceiving objects: 7% ( 5/71)\rReceiving objects: 8% ( 6/71)\rReceiving objects: 9% ( 7/71)\rReceiving objects: 11% ( 8/71)\rReceiving objects: 12% ( 9/71)\rReceiving objects: 14% (10/71)\rReceiving objects: 15% (11/71)\rReceiving objects: 16% (12/71)\rReceiving objects: 18% (13/71)\rReceiving objects: 19% (14/71)\rReceiving objects: 21% (15/71)\rReceiving objects: 22% (16/71)\rReceiving objects: 23% (17/71)\rReceiving objects: 25% (18/71)\rReceiving objects: 26% (19/71)\rReceiving objects: 28% (20/71)\rReceiving objects: 29% (21/71)\rReceiving objects: 30% (22/71)\rReceiving objects: 32% (23/71)\rReceiving objects: 33% (24/71)\rReceiving objects: 35% (25/71)\rReceiving objects: 36% (26/71)\rReceiving objects: 38% (27/71)\rReceiving objects: 39% (28/71)\rReceiving objects: 40% (29/71)\rReceiving objects: 42% (30/71)\rReceiving objects: 43% (31/71)\rReceiving objects: 45% (32/71)\rReceiving objects: 46% (33/71)\rReceiving objects: 47% (34/71)\rReceiving objects: 49% (35/71)\rReceiving objects: 50% (36/71)\rReceiving objects: 52% (37/71)\rReceiving objects: 53% (38/71)\rReceiving objects: 54% (39/71)\rReceiving objects: 56% (40/71)\rReceiving objects: 57% (41/71)\rReceiving objects: 59% (42/71)\rReceiving objects: 60% (43/71)\rReceiving objects: 61% (44/71)\rReceiving objects: 63% (45/71)\rReceiving objects: 64% (46/71)\rReceiving objects: 66% (47/71)\rReceiving objects: 67% (48/71)\rReceiving objects: 69% (49/71)\rReceiving objects: 70% (50/71)\rReceiving objects: 71% (51/71)\rReceiving objects: 73% (52/71)\rReceiving objects: 74% (53/71)\rReceiving objects: 76% (54/71)\rReceiving objects: 77% (55/71)\rReceiving objects: 78% (56/71)\rReceiving objects: 80% (57/71)\rReceiving objects: 81% (58/71)\rReceiving objects: 83% (59/71)\rReceiving objects: 84% (60/71)\rReceiving objects: 85% (61/71)\rReceiving objects: 87% (62/71)\rReceiving objects: 88% (63/71)\rReceiving objects: 90% (64/71)\rReceiving objects: 91% (65/71)\rReceiving objects: 92% (66/71)\rReceiving objects: 94% (67/71)\rReceiving objects: 95% (68/71)\rReceiving objects: 97% (69/71)\rReceiving objects: 98% (70/71)\rReceiving objects: 100% (71/71)\rReceiving objects: 100% (71/71)\n\rResolving deltas: 2% ( 1/37)\rResolving deltas: 5% ( 2/37)\rResolving deltas: 8% ( 3/37)\rResolving deltas: 10% ( 4/37)\rResolving deltas: 13% ( 5/37)\rResolving deltas: 16% ( 6/37)\rResolving deltas: 18% ( 7/37)\rResolving deltas: 21% ( 8/37)\rResolving deltas: 24% ( 9/37)\rResolving deltas: 27% (10/37)\rResolving deltas: 29% (11/37)\rResolving deltas: 32% (12/37)\rResolving deltas: 35% (13/37)\rResolving deltas: 37% (14/37)\rResolving deltas: 40% (15/37)\rResolving deltas: 43% (16/37)\rResolving deltas: 45% (17/37)\rResolving deltas: 48% (18/37)\rResolving deltas: 51% (19/37)\rResolving deltas: 54% (20/37)\rResolving deltas: 56% (21/37)\rResolving deltas: 59% (22/37)\rResolving deltas: 62% (23/37)\rResolving deltas: 64% (24/37)\rResolving deltas: 67% (25/37)\rResolving deltas: 70% (26/37)\rResolving deltas: 72% (27/37)\rResolving deltas: 75% (28/37)\rResolving deltas: 78% (29/37)\rResolving deltas: 81% (30/37)\rResolving deltas: 83% (31/37)\rResolving deltas: 86% (32/37)\rResolving deltas: 89% (33/37)\rResolving deltas: 91% (34/37)\rResolving deltas: 94% (35/37)\rResolving deltas: 97% (36/37)\rResolving deltas: 100% (37/37)\rResolving deltas: 100% (37/37)\n\rUpdating references: 100% (1/1)\rUpdating references: 100% (1/1)\nChecking out origin/master ...\n"
  :status :failure
  :commits
  [{:hash "aca198ae8fe5c47b9080be38dc3cd70217f26ea9"
    :msg "0"
    :author "Krzysztof Władyka "
    :timestamp #inst "2017-01-16T17:01:16.000-00:00"}]
  :outputs
  {(2 1 2)
   {:out
    "No old or current revision found.\nCurrent HEAD:\naca198ae8fe5c47b9080be38dc3cd70217f26ea9 | 2017-01-16 18:01:16 +0100 | Krzysztof Władyka | 0\n"
    :status :success
    :commits
    [{:hash "aca198ae8fe5c47b9080be38dc3cd70217f26ea9"
      :msg "0"
      :author "Krzysztof Władyka "
      :timestamp #inst "2017-01-16T17:01:16.000-00:00"}]}
   (3 1 2)
   {:exit 1
    :out
    "\nlein test foo.core-test\n\nlein test :only foo.core-test/a-test\n\nFAIL in (a-test) (core_test.clj:7)\nFIXME, I fail.\nexpected: 0\n actual: 1\n diff: - 0\n + 1\n\nRan 1 tests containing 1 assertions.\n1 failures, 0 errors.\nTests failed.\n"
    :status :failure}
   (1 1 2)
   {:out
    "Cloning git@github.com:kwladyka/lambdacd-test.git ...\n\rremote:Counting objects: 71\rremote:Counting objects: 71\n\rremote:Compressing objects: 2% ( 1/37)\rremote:Compressing objects: 5% ( 2/37)\rremote:Compressing objects: 8% ( 3/37)\rremote:Compressing objects: 10% ( 4/37)\rremote:Compressing objects: 13% ( 5/37)\rremote:Compressing objects: 16% ( 6/37)\rremote:Compressing objects: 18% ( 7/37)\rremote:Compressing objects: 21% ( 8/37)\rremote:Compressing objects: 24% ( 9/37)\rremote:Compressing objects: 27% (10/37)\rremote:Compressing objects: 29% (11/37)\rremote:Compressing objects: 32% (12/37)\rremote:Compressing objects: 35% (13/37)\rremote:Compressing objects: 37% (14/37)\rremote:Compressing objects: 40% (15/37)\rremote:Compressing objects: 43% (16/37)\rremote:Compressing objects: 45% (17/37)\rremote:Compressing objects: 48% (18/37)\rremote:Compressing objects: 51% (19/37)\rremote:Compressing objects: 54% (20/37)\rremote:Compressing objects: 56% (21/37)\rremote:Compressing objects: 59% (22/37)\rremote:Compressing objects: 62% (23/37)\rremote:Compressing objects: 64% (24/37)\rremote:Compressing objects: 67% (25/37)\rremote:Compressing objects: 70% (26/37)\rremote:Compressing objects: 72% (27/37)\rremote:Compressing objects: 75% (28/37)\rremote:Compressing objects: 78% (29/37)\rremote:Compressing objects: 81% (30/37)\rremote:Compressing objects: 83% (31/37)\rremote:Compressing objects: 86% (32/37)\rremote:Compressing objects: 89% (33/37)\rremote:Compressing objects: 91% (34/37)\rremote:Compressing objects: 94% (35/37)\rremote:Compressing objects: 97% (36/37)\rremote:Compressing objects: 100% (37/37)\rremote:Compressing objects: 100% (37/37)\n\rReceiving objects: 1% ( 1/71)\rReceiving objects: 2% ( 2/71)\rReceiving objects: 4% ( 3/71)\rReceiving objects: 5% ( 4/71)\rReceiving objects: 7% ( 5/71)\rReceiving objects: 8% ( 6/71)\rReceiving objects: 9% ( 7/71)\rReceiving objects: 11% ( 8/71)\rReceiving objects: 12% ( 9/71)\rReceiving objects: 14% (10/71)\rReceiving objects: 15% (11/71)\rReceiving objects: 16% (12/71)\rReceiving objects: 18% (13/71)\rReceiving objects: 19% (14/71)\rReceiving objects: 21% (15/71)\rReceiving objects: 22% (16/71)\rReceiving objects: 23% (17/71)\rReceiving objects: 25% (18/71)\rReceiving objects: 26% (19/71)\rReceiving objects: 28% (20/71)\rReceiving objects: 29% (21/71)\rReceiving objects: 30% (22/71)\rReceiving objects: 32% (23/71)\rReceiving objects: 33% (24/71)\rReceiving objects: 35% (25/71)\rReceiving objects: 36% (26/71)\rReceiving objects: 38% (27/71)\rReceiving objects: 39% (28/71)\rReceiving objects: 40% (29/71)\rReceiving objects: 42% (30/71)\rReceiving objects: 43% (31/71)\rReceiving objects: 45% (32/71)\rReceiving objects: 46% (33/71)\rReceiving objects: 47% (34/71)\rReceiving objects: 49% (35/71)\rReceiving objects: 50% (36/71)\rReceiving objects: 52% (37/71)\rReceiving objects: 53% (38/71)\rReceiving objects: 54% (39/71)\rReceiving objects: 56% (40/71)\rReceiving objects: 57% (41/71)\rReceiving objects: 59% (42/71)\rReceiving objects: 60% (43/71)\rReceiving objects: 61% (44/71)\rReceiving objects: 63% (45/71)\rReceiving objects: 64% (46/71)\rReceiving objects: 66% (47/71)\rReceiving objects: 67% (48/71)\rReceiving objects: 69% (49/71)\rReceiving objects: 70% (50/71)\rReceiving objects: 71% (51/71)\rReceiving objects: 73% (52/71)\rReceiving objects: 74% (53/71)\rReceiving objects: 76% (54/71)\rReceiving objects: 77% (55/71)\rReceiving objects: 78% (56/71)\rReceiving objects: 80% (57/71)\rReceiving objects: 81% (58/71)\rReceiving objects: 83% (59/71)\rReceiving objects: 84% (60/71)\rReceiving objects: 85% (61/71)\rReceiving objects: 87% (62/71)\rReceiving objects: 88% (63/71)\rReceiving objects: 90% (64/71)\rReceiving objects: 91% (65/71)\rReceiving objects: 92% (66/71)\rReceiving objects: 94% (67/71)\rReceiving objects: 95% (68/71)\rReceiving objects: 97% (69/71)\rReceiving objects: 98% (70/71)\rReceiving objects: 100% (71/71)\rReceiving objects: 100% (71/71)\n\rResolving deltas: 2% ( 1/37)\rResolving deltas: 5% ( 2/37)\rResolving deltas: 8% ( 3/37)\rResolving deltas: 10% ( 4/37)\rResolving deltas: 13% ( 5/37)\rResolving deltas: 16% ( 6/37)\rResolving deltas: 18% ( 7/37)\rResolving deltas: 21% ( 8/37)\rResolving deltas: 24% ( 9/37)\rResolving deltas: 27% (10/37)\rResolving deltas: 29% (11/37)\rResolving deltas: 32% (12/37)\rResolving deltas: 35% (13/37)\rResolving deltas: 37% (14/37)\rResolving deltas: 40% (15/37)\rResolving deltas: 43% (16/37)\rResolving deltas: 45% (17/37)\rResolving deltas: 48% (18/37)\rResolving deltas: 51% (19/37)\rResolving deltas: 54% (20/37)\rResolving deltas: 56% (21/37)\rResolving deltas: 59% (22/37)\rResolving deltas: 62% (23/37)\rResolving deltas: 64% (24/37)\rResolving deltas: 67% (25/37)\rResolving deltas: 70% (26/37)\rResolving deltas: 72% (27/37)\rResolving deltas: 75% (28/37)\rResolving deltas: 78% (29/37)\rResolving deltas: 81% (30/37)\rResolving deltas: 83% (31/37)\rResolving deltas: 86% (32/37)\rResolving deltas: 89% (33/37)\rResolving deltas: 91% (34/37)\rResolving deltas: 94% (35/37)\rResolving deltas: 97% (36/37)\rResolving deltas: 100% (37/37)\rResolving deltas: 100% (37/37)\n\rUpdating references: 100% (1/1)\rUpdating references: 100% (1/1)\nChecking out origin/master ...\n"
    :status :success}}}
 :rerun-for-retrigger false
 :build-number 8
 :step-id (1 2)}
```

slack message:  
```edn
{:final-result
 {:exit 1
  :global {}
  :out
  "Cloning git@github.com:kwladyka/lambdacd-test.git ...\n\rremote:Counting objects: 71\rremote:Counting objects: 71\n\rremote:Compressing objects:   2% ( 1/37)\rremote:Compressing objects:   5% ( 2/37)\rremote:Compressing objects:   8% ( 3/37)\rremote:Compressing objects:  10% ( 4/37)\rremote:Compressing objects:  13% ( 5/37)\rremote:Compressing objects:  16% ( 6/37)\rremote:Compressing objects:  18% ( 7/37)\rremote:Compressing objects:  21% ( 8/37)\rremote:Compressing objects:  24% ( 9/37)\rremote:Compressing objects:  27% (10/37)\rremote:Compressing objects:  29% (11/37)\rremote:Compressing objects:  32% (12/37)\rremote:Compressing objects:  35% (13/37)\rremote:Compressing objects:  37% (14/37)\rremote:Compressing objects:  40% (15/37)\rremote:Compressing objects:  43% (16/37)\rremote:Compressing objects:  45% (17/37)\rremote:Compressing objects:  48% (18/37)\rremote:Compressing objects:  51% (19/37)\rremote:Compressing objects:  54% (20/37)\rremote:Compressing objects:  56% (21/37)\rremote:Compressing objects:  59% (22/37)\rremote:Compressing objects:  62% (23/37)\rremote:Compressing objects:  64% (24/37)\rremote:Compressing objects:  67% (25/37)\rremote:Compressing objects:  70% (26/37)\rremote:Compressing objects:  72% (27/37)\rremote:Compressing objects:  75% (28/37)\rremote:Compressing objects:  78% (29/37)\rremote:Compressing objects:  81% (30/37)\rremote:Compressing objects:  83% (31/37)\rremote:Compressing objects:  86% (32/37)\rremote:Compressing objects:  89% (33/37)\rremote:Compressing objects:  91% (34/37)\rremote:Compressing objects:  94% (35/37)\rremote:Compressing objects:  97% (36/37)\rremote:Compressing objects: 100% (37/37)\rremote:Compressing objects: 100% (37/37)\n\rReceiving objects:        1% ( 1/71)\rReceiving objects:        2% ( 2/71)\rReceiving objects:        4% ( 3/71)\rReceiving objects:        5% ( 4/71)\rReceiving objects:        7% ( 5/71)\rReceiving objects:        8% ( 6/71)\rReceiving objects:        9% ( 7/71)\rReceiving objects:       11% ( 8/71)\rReceiving objects:       12% ( 9/71)\rReceiving objects:       14% (10/71)\rReceiving objects:       15% (11/71)\rReceiving objects:       16% (12/71)\rReceiving objects:       18% (13/71)\rReceiving objects:       19% (14/71)\rReceiving objects:       21% (15/71)\rReceiving objects:       22% (16/71)\rReceiving objects:       23% (17/71)\rReceiving objects:       25% (18/71)\rReceiving objects:       26% (19/71)\rReceiving objects:       28% (20/71)\rReceiving objects:       29% (21/71)\rReceiving objects:       30% (22/71)\rReceiving objects:       32% (23/71)\rReceiving objects:       33% (24/71)\rReceiving objects:       35% (25/71)\rReceiving objects:       36% (26/71)\rReceiving objects:       38% (27/71)\rReceiving objects:       39% (28/71)\rReceiving objects:       40% (29/71)\rReceiving objects:       42% (30/71)\rReceiving objects:       43% (31/71)\rReceiving objects:       45% (32/71)\rReceiving objects:       46% (33/71)\rReceiving objects:       47% (34/71)\rReceiving objects:       49% (35/71)\rReceiving objects:       50% (36/71)\rReceiving objects:       52% (37/71)\rReceiving objects:       53% (38/71)\rReceiving objects:       54% (39/71)\rReceiving objects:       56% (40/71)\rReceiving objects:       57% (41/71)\rReceiving objects:       59% (42/71)\rReceiving objects:       60% (43/71)\rReceiving objects:       61% (44/71)\rReceiving objects:       63% (45/71)\rReceiving objects:       64% (46/71)\rReceiving objects:       66% (47/71)\rReceiving objects:       67% (48/71)\rReceiving objects:       69% (49/71)\rReceiving objects:       70% (50/71)\rReceiving objects:       71% (51/71)\rReceiving objects:       73% (52/71)\rReceiving objects:       74% (53/71)\rReceiving objects:       76% (54/71)\rReceiving objects:       77% (55/71)\rReceiving objects:       78% (56/71)\rReceiving objects:       80% (57/71)\rReceiving objects:       81% (58/71)\rReceiving objects:       83% (59/71)\rReceiving objects:       84% (60/71)\rReceiving objects:       85% (61/71)\rReceiving objects:       87% (62/71)\rReceiving objects:       88% (63/71)\rReceiving objects:       90% (64/71)\rReceiving objects:       91% (65/71)\rReceiving objects:       92% (66/71)\rReceiving objects:       94% (67/71)\rReceiving objects:       95% (68/71)\rReceiving objects:       97% (69/71)\rReceiving objects:       98% (70/71)\rReceiving objects:      100% (71/71)\rReceiving objects:      100% (71/71)\n\rResolving deltas:         2% ( 1/37)\rResolving deltas:         5% ( 2/37)\rResolving deltas:         8% ( 3/37)\rResolving deltas:        10% ( 4/37)\rResolving deltas:        13% ( 5/37)\rResolving deltas:        16% ( 6/37)\rResolving deltas:        18% ( 7/37)\rResolving deltas:        21% ( 8/37)\rResolving deltas:        24% ( 9/37)\rResolving deltas:        27% (10/37)\rResolving deltas:        29% (11/37)\rResolving deltas:        32% (12/37)\rResolving deltas:        35% (13/37)\rResolving deltas:        37% (14/37)\rResolving deltas:        40% (15/37)\rResolving deltas:        43% (16/37)\rResolving deltas:        45% (17/37)\rResolving deltas:        48% (18/37)\rResolving deltas:        51% (19/37)\rResolving deltas:        54% (20/37)\rResolving deltas:        56% (21/37)\rResolving deltas:        59% (22/37)\rResolving deltas:        62% (23/37)\rResolving deltas:        64% (24/37)\rResolving deltas:        67% (25/37)\rResolving deltas:        70% (26/37)\rResolving deltas:        72% (27/37)\rResolving deltas:        75% (28/37)\rResolving deltas:        78% (29/37)\rResolving deltas:        81% (30/37)\rResolving deltas:        83% (31/37)\rResolving deltas:        86% (32/37)\rResolving deltas:        89% (33/37)\rResolving deltas:        91% (34/37)\rResolving deltas:        94% (35/37)\rResolving deltas:        97% (36/37)\rResolving deltas:       100% (37/37)\rResolving deltas:       100% (37/37)\n\rUpdating references:    100% (1/1)\rUpdating references:    100% (1/1)\nChecking out origin/master ...\n"
  :status :failure
  :commits
  [{:hash "aca198ae8fe5c47b9080be38dc3cd70217f26ea9"
    :msg "0"
    :author "Krzysztof Władyka <hidden>"
    :timestamp #inst "2017-01-16T17:01:16.000-00:00"}]
  :outputs
  {(1 2)
   {:exit 1
    :global {}
    :out
    "Cloning git@github.com:kwladyka/lambdacd-test.git ...\n\rremote:Counting objects: 71\rremote:Counting objects: 71\n\rremote:Compressing objects:   2% ( 1/37)\rremote:Compressing objects:   5% ( 2/37)\rremote:Compressing objects:   8% ( 3/37)\rremote:Compressing objects:  10% ( 4/37)\rremote:Compressing objects:  13% ( 5/37)\rremote:Compressing objects:  16% ( 6/37)\rremote:Compressing objects:  18% ( 7/37)\rremote:Compressing objects:  21% ( 8/37)\rremote:Compressing objects:  24% ( 9/37)\rremote:Compressing objects:  27% (10/37)\rremote:Compressing objects:  29% (11/37)\rremote:Compressing objects:  32% (12/37)\rremote:Compressing objects:  35% (13/37)\rremote:Compressing objects:  37% (14/37)\rremote:Compressing objects:  40% (15/37)\rremote:Compressing objects:  43% (16/37)\rremote:Compressing objects:  45% (17/37)\rremote:Compressing objects:  48% (18/37)\rremote:Compressing objects:  51% (19/37)\rremote:Compressing objects:  54% (20/37)\rremote:Compressing objects:  56% (21/37)\rremote:Compressing objects:  59% (22/37)\rremote:Compressing objects:  62% (23/37)\rremote:Compressing objects:  64% (24/37)\rremote:Compressing objects:  67% (25/37)\rremote:Compressing objects:  70% (26/37)\rremote:Compressing objects:  72% (27/37)\rremote:Compressing objects:  75% (28/37)\rremote:Compressing objects:  78% (29/37)\rremote:Compressing objects:  81% (30/37)\rremote:Compressing objects:  83% (31/37)\rremote:Compressing objects:  86% (32/37)\rremote:Compressing objects:  89% (33/37)\rremote:Compressing objects:  91% (34/37)\rremote:Compressing objects:  94% (35/37)\rremote:Compressing objects:  97% (36/37)\rremote:Compressing objects: 100% (37/37)\rremote:Compressing objects: 100% (37/37)\n\rReceiving objects:        1% ( 1/71)\rReceiving objects:        2% ( 2/71)\rReceiving objects:        4% ( 3/71)\rReceiving objects:        5% ( 4/71)\rReceiving objects:        7% ( 5/71)\rReceiving objects:        8% ( 6/71)\rReceiving objects:        9% ( 7/71)\rReceiving objects:       11% ( 8/71)\rReceiving objects:       12% ( 9/71)\rReceiving objects:       14% (10/71)\rReceiving objects:       15% (11/71)\rReceiving objects:       16% (12/71)\rReceiving objects:       18% (13/71)\rReceiving objects:       19% (14/71)\rReceiving objects:       21% (15/71)\rReceiving objects:       22% (16/71)\rReceiving objects:       23% (17/71)\rReceiving objects:       25% (18/71)\rReceiving objects:       26% (19/71)\rReceiving objects:       28% (20/71)\rReceiving objects:       29% (21/71)\rReceiving objects:       30% (22/71)\rReceiving objects:       32% (23/71)\rReceiving objects:       33% (24/71)\rReceiving objects:       35% (25/71)\rReceiving objects:       36% (26/71)\rReceiving objects:       38% (27/71)\rReceiving objects:       39% (28/71)\rReceiving objects:       40% (29/71)\rReceiving objects:       42% (30/71)\rReceiving objects:       43% (31/71)\rReceiving objects:       45% (32/71)\rReceiving objects:       46% (33/71)\rReceiving objects:       47% (34/71)\rReceiving objects:       49% (35/71)\rReceiving objects:       50% (36/71)\rReceiving objects:       52% (37/71)\rReceiving objects:       53% (38/71)\rReceiving objects:       54% (39/71)\rReceiving objects:       56% (40/71)\rReceiving objects:       57% (41/71)\rReceiving objects:       59% (42/71)\rReceiving objects:       60% (43/71)\rReceiving objects:       61% (44/71)\rReceiving objects:       63% (45/71)\rReceiving objects:       64% (46/71)\rReceiving objects:       66% (47/71)\rReceiving objects:       67% (48/71)\rReceiving objects:       69% (49/71)\rReceiving objects:       70% (50/71)\rReceiving objects:       71% (51/71)\rReceiving objects:       73% (52/71)\rReceiving objects:       74% (53/71)\rReceiving objects:       76% (54/71)\rReceiving objects:       77% (55/71)\rReceiving objects:       78% (56/71)\rReceiving objects:       80% (57/71)\rReceiving objects:       81% (58/71)\rReceiving objects:       83% (59/71)\rReceiving objects:       84% (60/71)\rReceiving objects:       85% (61/71)\rReceiving objects:       87% (62/71)\rReceiving objects:       88% (63/71)\rReceiving objects:       90% (64/71)\rReceiving objects:       91% (65/71)\rReceiving objects:       92% (66/71)\rReceiving objects:       94% (67/71)\rReceiving objects:       95% (68/71)\rReceiving objects:       97% (69/71)\rReceiving objects:       98% (70/71)\rReceiving objects:      100% (71/71)\rReceiving objects:      100% (71/71)\n\rResolving deltas:         2% ( 1/37)\rResolving deltas:         5% ( 2/37)\rResolving deltas:         8% ( 3/37)\rResolving deltas:        10% ( 4/37)\rResolving deltas:        13% ( 5/37)\rResolving deltas:        16% ( 6/37)\rResolving deltas:        18% ( 7/37)\rResolving deltas:        21% ( 8/37)\rResolving deltas:        24% ( 9/37)\rResolving deltas:        27% (10/37)\rResolving deltas:        29% (11/37)\rResolving deltas:        32% (12/37)\rResolving deltas:        35% (13/37)\rResolving deltas:        37% (14/37)\rResolving deltas:        40% (15/37)\rResolving deltas:        43% (16/37)\rResolving deltas:        45% (17/37)\rResolving deltas:        48% (18/37)\rResolving deltas:        51% (19/37)\rResolving deltas:        54% (20/37)\rResolving deltas:        56% (21/37)\rResolving deltas:        59% (22/37)\rResolving deltas:        62% (23/37)\rResolving deltas:        64% (24/37)\rResolving deltas:        67% (25/37)\rResolving deltas:        70% (26/37)\rResolving deltas:        72% (27/37)\rResolving deltas:        75% (28/37)\rResolving deltas:        78% (29/37)\rResolving deltas:        81% (30/37)\rResolving deltas:        83% (31/37)\rResolving deltas:        86% (32/37)\rResolving deltas:        89% (33/37)\rResolving deltas:        91% (34/37)\rResolving deltas:        94% (35/37)\rResolving deltas:        97% (36/37)\rResolving deltas:       100% (37/37)\rResolving deltas:       100% (37/37)\n\rUpdating references:    100% (1/1)\rUpdating references:    100% (1/1)\nChecking out origin/master ...\n"
    :status :failure
    :commits
    [{:hash "aca198ae8fe5c47b9080be38dc3cd70217f26ea9"
      :msg "0"
      :author "Krzysztof Władyka <hidden>"
      :timestamp #inst "2017-01-16T17:01:16.000-00:00"}]
    :outputs
    {(2 1 2)
     {:out
      "No old or current revision found.\nCurrent HEAD:\naca198ae8fe5c47b9080be38dc3cd70217f26ea9 | 2017-01-16 18:01:16 +0100 | Krzysztof Władyka <hidden> | 0\n"
      :status :success
      :commits
      [{:hash "aca198ae8fe5c47b9080be38dc3cd70217f26ea9"
        :msg "0"
        :author "Krzysztof Władyka <hidden>"
        :timestamp #inst "2017-01-16T17:01:16.000-00:00"}]}
     (3 1 2)
     {:exit 1
      :out
      "\nlein test foo.core-test\n\nlein test :only foo.core-test/a-test\n\nFAIL in (a-test) (core_test.clj:7)\nFIXME, I fail.\nexpected: 0\n  actual: 1\n    diff: - 0\n          + 1\n\nRan 1 tests containing 1 assertions.\n1 failures, 0 errors.\nTests failed.\n"
      :status :failure}
     (1 1 2)
     {:out
      "Cloning git@github.com:kwladyka/lambdacd-test.git ...\n\rremote:Counting objects: 71\rremote:Counting objects: 71\n\rremote:Compressing objects:   2% ( 1/37)\rremote:Compressing objects:   5% ( 2/37)\rremote:Compressing objects:   8% ( 3/37)\rremote:Compressing objects:  10% ( 4/37)\rremote:Compressing objects:  13% ( 5/37)\rremote:Compressing objects:  16% ( 6/37)\rremote:Compressing objects:  18% ( 7/37)\rremote:Compressing objects:  21% ( 8/37)\rremote:Compressing objects:  24% ( 9/37)\rremote:Compressing objects:  27% (10/37)\rremote:Compressing objects:  29% (11/37)\rremote:Compressing objects:  32% (12/37)\rremote:Compressing objects:  35% (13/37)\rremote:Compressing objects:  37% (14/37)\rremote:Compressing objects:  40% (15/37)\rremote:Compressing objects:  43% (16/37)\rremote:Compressing objects:  45% (17/37)\rremote:Compressing objects:  48% (18/37)\rremote:Compressing objects:  51% (19/37)\rremote:Compressing objects:  54% (20/37)\rremote:Compressing objects:  56% (21/37)\rremote:Compressing objects:  59% (22/37)\rremote:Compressing objects:  62% (23/37)\rremote:Compressing objects:  64% (24/37)\rremote:Compressing objects:  67% (25/37)\rremote:Compressing objects:  70% (26/37)\rremote:Compressing objects:  72% (27/37)\rremote:Compressing objects:  75% (28/37)\rremote:Compressing objects:  78% (29/37)\rremote:Compressing objects:  81% (30/37)\rremote:Compressing objects:  83% (31/37)\rremote:Compressing objects:  86% (32/37)\rremote:Compressing objects:  89% (33/37)\rremote:Compressing objects:  91% (34/37)\rremote:Compressing objects:  94% (35/37)\rremote:Compressing objects:  97% (36/37)\rremote:Compressing objects: 100% (37/37)\rremote:Compressing objects: 100% (37/37)\n\rReceiving objects:        1% ( 1/71)\rReceiving objects:        2% ( 2/71)\rReceiving objects:        4% ( 3/71)\rReceiving objects:        5% ( 4/71)\rReceiving objects:        7% ( 5/71)\rReceiving objects:        8% ( 6/71)\rReceiving objects:        9% ( 7/71)\rReceiving objects:       11% ( 8/71)\rReceiving objects:       12% ( 9/71)\rReceiving objects:       14% (10/71)\rReceiving objects:       15% (11/71)\rReceiving objects:       16% (12/71)\rReceiving objects:       18% (13/71)\rReceiving objects:       19% (14/71)\rReceiving objects:       21% (15/71)\rReceiving objects:       22% (16/71)\rReceiving objects:       23% (17/71)\rReceiving objects:       25% (18/71)\rReceiving objects:       26% (19/71)\rReceiving objects:       28% (20/71)\rReceiving objects:       29% (21/71)\rReceiving objects:       30% (22/71)\rReceiving objects:       32% (23/71)\rReceiving objects:       33% (24/71)\rReceiving objects:       35% (25/71)\rReceiving objects:       36% (26/71)\rReceiving objects:       38% (27/71)\rReceiving objects:       39% (28/71)\rReceiving objects:       40% (29/71)\rReceiving objects:       42% (30/71)\rReceiving objects:       43% (31/71)\rReceiving objects:       45% (32/71)\rReceiving objects:       46% (33/71)\rReceiving objects:       47% (34/71)\rReceiving objects:       49% (35/71)\rReceiving objects:       50% (36/71)\rReceiving objects:       52% (37/71)\rReceiving objects:       53% (38/71)\rReceiving objects:       54% (39/71)\rReceiving objects:       56% (40/71)\rReceiving objects:       57% (41/71)\rReceiving objects:       59% (42/71)\rReceiving objects:       60% (43/71)\rReceiving objects:       61% (44/71)\rReceiving objects:       63% (45/71)\rReceiving objects:       64% (46/71)\rReceiving objects:       66% (47/71)\rReceiving objects:       67% (48/71)\rReceiving objects:       69% (49/71)\rReceiving objects:       70% (50/71)\rReceiving objects:       71% (51/71)\rReceiving objects:       73% (52/71)\rReceiving objects:       74% (53/71)\rReceiving objects:       76% (54/71)\rReceiving objects:       77% (55/71)\rReceiving objects:       78% (56/71)\rReceiving objects:       80% (57/71)\rReceiving objects:       81% (58/71)\rReceiving objects:       83% (59/71)\rReceiving objects:       84% (60/71)\rReceiving objects:       85% (61/71)\rReceiving objects:       87% (62/71)\rReceiving objects:       88% (63/71)\rReceiving objects:       90% (64/71)\rReceiving objects:       91% (65/71)\rReceiving objects:       92% (66/71)\rReceiving objects:       94% (67/71)\rReceiving objects:       95% (68/71)\rReceiving objects:       97% (69/71)\rReceiving objects:       98% (70/71)\rReceiving objects:      100% (71/71)\rReceiving objects:      100% (71/71)\n\rResolving deltas:         2% ( 1/37)\rResolving deltas:         5% ( 2/37)\rResolving deltas:         8% ( 3/37)\rResolving deltas:        10% ( 4/37)\rResolving deltas:        13% ( 5/37)\rResolving deltas:        16% ( 6/37)\rResolving deltas:        18% ( 7/37)\rResolving deltas:        21% ( 8/37)\rResolving deltas:        24% ( 9/37)\rResolving deltas:        27% (10/37)\rResolving deltas:        29% (11/37)\rResolving deltas:        32% (12/37)\rResolving deltas:        35% (13/37)\rResolving deltas:        37% (14/37)\rResolving deltas:        40% (15/37)\rResolving deltas:        43% (16/37)\rResolving deltas:        45% (17/37)\rResolving deltas:        48% (18/37)\rResolving deltas:        51% (19/37)\rResolving deltas:        54% (20/37)\rResolving deltas:        56% (21/37)\rResolving deltas:        59% (22/37)\rResolving deltas:        62% (23/37)\rResolving deltas:        64% (24/37)\rResolving deltas:        67% (25/37)\rResolving deltas:        70% (26/37)\rResolving deltas:        72% (27/37)\rResolving deltas:        75% (28/37)\rResolving deltas:        78% (29/37)\rResolving deltas:        81% (30/37)\rResolving deltas:        83% (31/37)\rResolving deltas:        86% (32/37)\rResolving deltas:        89% (33/37)\rResolving deltas:        91% (34/37)\rResolving deltas:        94% (35/37)\rResolving deltas:        97% (36/37)\rResolving deltas:       100% (37/37)\rResolving deltas:       100% (37/37)\n\rUpdating references:    100% (1/1)\rUpdating references:    100% (1/1)\nChecking out origin/master ...\n"
      :status :success}}}}}
 :rerun-for-retrigger false
 :build-number 8
 :step-id (2)}
```

**Comment by flosell** [comment]
To me, this looks like intended behavior. I pretty-printed the data, hopefully this makes it a bit more visible what's happening: 

Your pipeline is nested: `with-workspace` is a step that manages three child-steps and (not shown in your snippet but guessing from the data) it in turn is nested into something as well. 
Nesting steps "inherit" the status of their children in a way they see fit (for example, that's the main difference between `either` and `in-parallel`: one fails if one of its children fails, the other only fails if all children fail). 

This is why you see multiple events: One for the step that's actually failing and one for each parent that fails because of a failed child, propagating the failure to stop the pipeline. It's visible in the `:final-result`: Parent steps contain their childrens results in `:outputs` as a map step-id to step-result: 

In your example, the first message has the `:step-id` `(3 1 2)` ("third child of step `(1 2)`). That's the `lein-test` step. You'll find the same step-result again in the two following messages, only this time inside the `:outputs` map. 

Hope this helps as background. 

So what can do about this? Looks like you are only interested in the "leaf" steps that don't have children and do the actual work. Those should not have `:outputs` keys in their result so filtering for that should solve your problem 95%. 

You might skip over edge-cases this way though, e.g. you would not notice if the `with-workspace`-step failed because it is out of diskspace. If you want to catch such things, you'd probably need to make your filtering a bit smarter, e.g. by looking deeper into the `:outputs` map or keeping track of which other failures you already received.

**Comment by kwladyka** [comment]
Hmm, but on the other hand it provides misleading data. Because i got `:failure` on child steps which really `:success`. The data don't let to determine what really failed. I think this is weak point. The UI know which step really failed, how?

That is really strange people don't want such reporting... maybe they just didn't ask. Or i am strange :)

Alternative solution is message "build failed, click on [link to lambdacd pipeline] and check what". But i don't see event for the final `:failure` of build. In that case i want send this message only once for the build.

**Comment by flosell** [comment]
I'm not sure I follow you, can you provide an example that's misleading? The way I see it, you get a failure for step `(3 1 2)`, thats your `lein-test` that's failing and you get successes for `(1 1 2)` and `(2 1 2)` (those are `(steps/clone sb-repo)` and `git/list-changes`). That's the data the UI uses to show which steps failed. 

There is currently no event that reports the result of the whole pipeline-build (though I can see that this might be useful). For the moment, you could detect the pipeline failure by checking for failure events with a step-id of length 1 (a bit hacky though...)

**Comment by kwladyka** [comment]
<img width="858" alt="screen shot 2017-01-17 at 09 32 27" src="https://cloud.githubusercontent.com/assets/3903726/22013145/e33c6db6-dc97-11e6-825a-3e30768cdaab.png">
I am sorry i didn't clarify it on the beginning enough.

On the end i can check if `:outputs` exist and send messsge if not, but you said it wouldn't cover 100% cases. Alternative i can check if `:steps-id` length is 1 and then inside `:outputs` looking `:failure`. Will it cover 100%?

**Comment by flosell** [comment]
Ah, ok. LambdaCD stops the pipeline if one of the root-level steps (those with step-id with length 1) has `:status != :success` so filtering for those events will tell you if the complete build stopped (and you can drill into the step results for details). If that helps, I wouldn't mind adding `:pipeline-start`, `:pipeline-stopped` events to simplify this. If this is all you are interested in, that's probably the approach to go for and should cover everything you need. 

If you are interested in nested failures that don't stop the pipeline (e.g. an `either` with a failing child), you probably need a bit of logic to aggregate and filter for what you are interested in.

**Comment by kwladyka** [comment]
I am interesting in failures which stop the whole pipeline. So if primary step fail i want to send message what happened. First i didn't thought about parallel cases, but you are right. In such situations i need to know about every failed steps. This is not only one step then.

Just on the end i want to know why process failed and i want to know that only once (determine which events should i process).

If you can add this events it will be easier. I am still confuse how to do it in the right way now.

**Comment by flosell** [comment]
Hi @kwladyka, sorry for taking a while. I actually looked at your use-case in a bit more detail and added a few features that you'd probably need in implementing this. They will all be part of the next LambdaCD release (which will likely come soon). 

Here's what's new: 
* `:pipeline-finished` event
* `lambdacd.steps.result/flatten-step-result-outputs` to convert the nested `:outputs` structure you receive from `:pipeline-finished` into a map that contains all step ids so you can easily filter for failed steps
* `lambdacd.presentation.pipeline-structure/step-display-representation-by-step-id` to get information about a build step (e.g. name) for a given step-id. Should come in handy if you want a human-readable way of telling users which step failed

Also, here's an example of how to combine all this into a notification if a build fails: https://gist.github.com/flosell/ff9976e9744814d60fcbb7d06f488f71
It waits for `:pipeline-finished` events, looks up all failed steps, excludes all failed steps that are parents of another failed step and then looks up and prints name and step results for those remaining. 
Probably not perfect yet but a nice example to get you started. I also wouldn't mind collaborating on this if you plan to build such functionality as a separate open-source library.

**Comment by flosell** [comment]
Just [released](https://github.com/flosell/lambdacd/releases/tag/0.13.0) the improvements mentioned above. 

@kwladyka Does this help?

**Comment by kwladyka** [comment]
Yes. It helps. Great.

I use this one trigger to send to slack information about `:failed` and `:success`. What is important remember to filter `:killed` to not send logs about restarting the app.

In the future it could be nice to push `failed-steps` into lambdacd code to easy get failed steps with one function.

**Comment by flosell** [comment]
Thanks for the feedback! I created a new issue for your idea (feel free to comment and add more detail) and I'm closing this one for now.
